packages/marshal/src/encodeToSmallcaps.js

```diff
@@ -0,0 +1,452 @@
+        }
+        if (is(passable, -0)) {
+          return 0;
+        }
+        if (passable === Infinity) {
+          return '#Infinity';
+        }
+        if (passable === -Infinity) {
+          return '#-Infinity';
+        }
+        // All other numbers pass through to JSON
+        return passable;
+      }
+      case 'bigint': {
+        const str = String(passable);
+        return passable < 0n ? str : `+${str}`;
+      }
+      case 'copyRecord': {
+        // Currently copyRecord allows only string keys so this will
+        // work. If we allow sortable symbol keys, this will need to
+        // become more interesting.
+        const names = ownKeys(passable).sort();
+        return fromEntries(
+          names.map(name => [
+            encodeToSmallcapsRecur(name),
+            encodeToSmallcapsRecur(passable[name]),
+          ]),
+        );
+      }
+      case 'copyArray': {
+        return passable.map(encodeToSmallcapsRecur);
+      }
+      case 'tagged': {
+        return {
+          '#tag': getTag(passable),
+          payload: encodeToSmallcapsRecur(passable.payload),
+        };
+      }
+      case 'remotable': {
+        const result = encodeRemotableToSmallcaps(passable);
+        if (typeof result === 'string' && result.startsWith('$')) {
+          return result;
+        }
+        assert.fail(
+          X`internal: Remotable encoding must start with "$": ${result}`,
+        );
+      }
+      case 'promise': {
+        const result = encodePromiseToSmallcaps(passable);
+        if (typeof result === 'string' && result.startsWith('&')) {
+          return result;
+        }
+        assert.fail(
+          X`internal: Promise encoding must start with "&": ${result}`,
+        );
+      }
+      case 'error': {
+        const result = encodeErrorToSmallcaps(passable);
+        if (
+          typeof result === 'object' &&
+          typeof result['#error'] === 'string'
+        ) {
+          return result;
+        }
+        assert.fail(
+          X`internal: Error encoding must have "#error" property: ${result}`,
+        );
+      }
+      default: {
+        assert.fail(X`unrecognized passStyle ${q(passStyle)}`, TypeError);
+      }
+    }
+  };
+  const encodeToSmallcaps = passable => {
+    if (ErrorHelper.canBeValid(passable, x => x)) {
+      // We pull out this special case to accommodate errors that are not
+      // valid Passables. For example, because they're not frozen.
+      // The special case can only ever apply at the root, and therefore
+      // outside the recursion, since an error could only be deeper in
+      // a passable structure if it were passable.
+      //
+      // We pull out this special case because, for these errors, we're much
+      // more interested in reporting whatever diagnostic information they
+      // carry than we are about reporting problems encountered in reporting
+      // this information.
+      return harden(encodeErrorToSmallcaps(passable));
```

Comment:
```suggestion
      const result = encodeErrorToSmallcaps(passable);
      if (
        typeof result === 'object' &&
        hasOwnPropertyOf(result, '#error') &&
        typeof result['#error'] === 'string'
      ) {
        return harden(result);
      }
      assert.fail(
        X`internal: Error encoding must have "#error" property: ${result}`,
      );
```
Or alternatively and probably better, `encodeErrorToSmallcaps` and the other input encoders could be replaced up front with a wrapper that performs the assertions:
```patch
 export const makeEncodeToSmallcaps = ({
   encodeRemotableToSmallcaps = dontEncodeRemotableToSmallcaps,
   encodePromiseToSmallcaps = dontEncodePromiseToSmallcaps,
   encodeErrorToSmallcaps = dontEncodeErrorToSmallcaps,
 } = {}) => {
+  encodeRemotableToSmallcaps = enforceGuards(
+    encodeRemotableToSmallcaps,
+    // TODO? M.startsWith/M.matchesRegExp/etc.
+    M.call(M.any()).returns(M.and(M.string(), {length: M.gte(1), 0: '$'})),
+  );
+  encodePromiseToSmallcaps = enforceGuards(
+    encodePromiseToSmallcaps,
+    M.call(M.any()).returns(M.and(M.string(), {length: M.gte(1), 0: '&'})),
+  );
+  encodeErrorToSmallcaps = enforceGuards(
+    encodeErrorToSmallcaps,
+    M.call(M.any()).returns({'#error': M.string()}),
+  );
```

Reply:
Did the first simple suggestion.

A fly in the ointment of the second suggestion is that `M` is defined by `@agoric/store`, and so difficult for marshal to import ;). Nevertheless, I agree that something along those lines would be better. But not doing it in this PR.